modern-publish-plugin.md

```diff
@@ -0,0 +1,222 @@
+
+## New Structure
+
+For compatibility purposes the current `com.vanniktech.maven.publish` plugin
+behavior is kept as it is, with the exception of APIs that we decide to remove. 
+Most of the goals are solved through a set of new plugins for specific 
+platforms. These plugins provide APIs that replace the magic setup mechanisms
+of the current plugin. To avoid having to maintain two implentations the old
+plugin will internally just apply the new plugins and use their public APIs. 
+Over time we can decide if `com.vanniktech.maven.publish` should be deprecated
+or removed.
+
+### Plugins
+
+Specific plugins for each supported project type:
+- `com.vanniktech.maven.publish.java`
+- `com.vanniktech.maven.publish.gradle`
+- `com.vanniktech.maven.publish.android`
+- `com.vanniktech.maven.publish.kotlin.jvm`
+- `com.vanniktech.maven.publish.kotlin.js`
+- `com.vanniktech.maven.publish.kotlin.mpp`
+
+The main motivation for separating these, is to provide specific APIs for each project
+without running into race conditions when something like `plugins.withId` is used or 
+requiring ordering. Another reason is that it allows us to nicely separate the code. 
+
+On top of that there will be a `com.vanniktech.maven.publish.base` plugin that 
+gets applied by all of these and the existing `com.vanniktech.maven.publish` as 
+described above.
+
+#### `com.vanniktech.maven.publish.base`
+
+- applies `maven-publish` Gradle plugin
+- provides the API that is shared for all project types
+- can be used for shared configurations using `plugins.withId(...)` in projects with mixed types
+
+```
+mavenPublish {
+  # applies signing plugin and tells Gradle to sign all publications
+  signAllPublications()
+  # adds the Sonatype release/snapshot repo as targets based on version
+  # uses Gradle credentials API for username/password
+  # automatically calls signAllPublications()
+  # sets nexus.baseUrl, nexus.respositoryUserName, nexus.respositoryPassword
+  publishToMavenCentral()
+
+  # Same API as Gradle's, can probably just use their interfaces/classes
+  # Will be applied to all publications
+  pom {
+    name = "My Library"
+    description = "A description of what my library does."
+    inceptionYear = "2020"
+    url = "https://github.com/username/mylibrary/"
+    
+    scm {
+        connection = "scm:git:git://github.com/username/mylibrary.git"
+        developerConnection = "scm:git:ssh://git@github.com/username/mylibrary.git"
+        url = "https://github.com/username/mylibrary/"
+    } 
+    
+    licenses {
+        license {
+            name = "The Apache Software License, Version 2.0"
+            url = "https://www.apache.org/licenses/LICENSE-2.0.txt"
+            distribution = "repo"
+        }
+    }
+    developers {
+        developer {
+            id = "username"
+            name = "User Name"
+            url = "https://github.com/username/"
+        }
+    }
+  }
+
+  # Configure closeAndReleaseRepository
+  nexus {
```

Comment:
What about merging this with `publishToMavenCentral` ? Both of them require the sonatype username/password and this would allow to do fancy stuff like ensuring everything gets uploaded to the same staging repo:

```
addNexusRepository(
  baseUrl = ..
  username = ..
  password = ..
  stagingProfile = .. // if null will use the first stagingProfile
  repositoryId = .. // if null will create a new staging repository
)

addMavenCentralRepository() // a shortcut for addNexusRepository with baseUrl already configured
```

Reply:
At the very least `publishToMavenCentral` will configure nexus: https://github.com/vanniktech/gradle-maven-publish-plugin/pull/173/files#diff-6a981871304a5f63c9fda36a805d124700fa5632c907f1c8e50bd91303863c16R53 So you wouldn't need to call it yourself unless you want to set a `stagingProfile`, but maybe that can be an optional parameter.

When I started writing the code for the base extension, I was unsure about how to handle `nexus`. In general I don't want to duplicate a perfectly fine Gradle API again (adding repositories). I also don't know how much of a use case `closeAndReleaseRepository` has outside of Sonatype OSSRH. Does someone who self hosts a Nexus instance use staging repositories for publishing or would they just publish directly to the repository. Maybe we leave out the API for now on base and see if someone needs it.

Reply:
>  I also don't know how much of a use case closeAndReleaseRepository has outside of Sonatype OSSRH. 

I guess OSSRH is the main use case.

> So you wouldn't need to call it yourself unless you want to set a stagingProfile, but maybe that can be an optional parameter.

Yup, something like `publishToMavenCentral(stagingProfile: String? = null)`

I think there is a lot of value in the pluging avoiding split staging repositories though, i.e having `publishToMavenCentral` declare a repository with `url=`https://oss.sonatype.org/service/local/staging/deployByRepositoryId/$repositoryId`. This has been a pain point for many users.

All in all, something like this:

```
// adds a new ossrh repository that creates a new staging repository for this build
// @param stagingProfile: the staging profile to use. If null, will use the first one or fail if there is more than one
// registers a closeAndReleaseRepository task that will close and release the newly created staging repository (or fail if more than 1 repository are open)
publishToMavenCentral(stagingProfile: String? = null)
```

We can refine later by having the repositoryId an output property so that it can be reused. But avoiding the split repositories would be a nice first step.

Reply:
Would we even need `stagingProfile` if we create the staging repository through the API? The API seems to require a profile id which would be something different. So maybe it makes sense to leave this out for now.

Reply:
Sorry I meant `stagingProfileId`, not `stagingProfile`, these are the same thing that you can see at https://oss.sonatype.org/#stagingProfiles;$stagingProfileId

The more I think about this, the more I think 2 versions would be nice actually 😅 :

```
// Creates a new staging repo using the given stagingProfileId or the first one found if null
publishToMavenCentral(stagingProfileId: String? = null)
// Does not create a new staging repo but reuses one that is already created
publishToMavenCentralWithStagingRepository(stagingRepositoryId: String)
```

I'd like to make a small Github action that creates a staging repository. If that works, using the 2nd version up there would allow to have all artifacts uploaded to the same staging repository, for an exemple in MPP  builds spread on multiple VMs.